analyzers/src/SonarAnalyzer.CSharp/Rules/MemberInitializerRedundant.RoslynCfg.cs

```diff
@@ -0,0 +1,111 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.CFG.Roslyn;
+using SonarAnalyzer.Extensions;
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    public sealed partial class MemberInitializerRedundant
+    {
+        private class RoslynChecker : CfgAllPathValidator
+        {
+            private readonly ISymbol memberToCheck;
+            private readonly ControlFlowGraph cfg;
+
+            public RoslynChecker(ControlFlowGraph cfg, ISymbol memberToCheck) : base(cfg)
+            {
+                this.cfg = cfg;
+                this.memberToCheck = memberToCheck;
+            }
+
+            // Returns true if the block contains assignment before access
+            protected override bool IsValid(BasicBlock block) =>
+                Match(block, checkReadBeforeWrite: false);
+
+            // Returns true if the block contains access before assignment
+            protected override bool IsInvalid(BasicBlock block) =>
+                Match(block, checkReadBeforeWrite: true);
+
+            private bool Match(BasicBlock block, bool checkReadBeforeWrite)
+            {
+                foreach (var operation in block.OperationsAndBranchValue)
+                {
+                    if (ProcessOperation(operation, cfg, checkReadBeforeWrite, out var result))
+                    {
+                        return result;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool ProcessOperation(IOperation operation, ControlFlowGraph controlFlowGraph, bool isRead, out bool result)
+            {
+                foreach (var child in operation.DescendantsAndSelf().ToReversedExecutionOrder())
```

Comment:
I see here the same problem as on the other PR (where this code got deleted) - https://github.com/SonarSource/sonar-dotnet/pull/4855#discussion_r709968524

>The `ToReversedExectuionOrder()` will descent into children as well so we would end up iterating all of them twice.
>
>The intended usage is like this. I'd be also fine with creating a handy overload of the extension method so we avoid this pitfall in the future as well.
```suggestion
                        foreach (var child in new[] {operation}.ToReversedExecutionOrder())
```
>Or maybe something like this one level higher, but that might change the semantics.
```
foreach (var operation in block.OperationsAndBranchValue.ToReversedExecutionOrder())
```
>or
```
foreach (var operation in block.OperationsAndBranchValue.Reverse().ToReversedExecutionOrder())
```

Reply:
duly noted on the other PR to take action on it before merging it

Reply:
For now, converted to array of one element

Reply:
Resolving - this follows in the other comment below https://github.com/SonarSource/sonar-dotnet/pull/4882#discussion_r713020235